### DISCUSSION TITLE
WAS와 웹서버의 차이를 설명해주세요.

### DISCUSSION THREAD

**Discussion by J-I-H-O** [2024-woowacourse-study/level-interview#154] [Network]
대답 못하면 테드한테 혼남

**Comment by nayonsoso** [comment]
### WebServer
- 정적(static) 컨텐츠를 제공하는 프로그램
- 클라이언트로부터 HTTP 요청을 받을 수 있다.
- 정적 컨텍츠 요청을 받을 시, 정적 컨텐츠를 응답한다.
- 동적 컨텐츠 요청을 받을 시, 이를 WAS로 전달하고, WAS의 응답을 클라이언트에 전달한다.
- 대표적인 WebServer 로는 Nginx 가 있다.

### WAS
- 동적(dynamic) 컨텐츠를 제공하는 프로그램
- 하지만 실제로는 ❗️대부분의 WAS는 WebServer 를 내장하고 있기 때문에, 정적 컨텐츠와 동적 컨텐츠를 모두 제공할 수 있다❗️
- 대표적인 WAS 로는 Tomcat이 있다.

### WAS가 WebServer를 내장하고 있다면서 왜 Nginx 같은 WebServer 가 필요한가?🧐
- 뭐 이유가 있다곤 하는데 지금은 그닥 와닿진 않는다..

**Comment by gyungchan-jo** [comment]
웹서버(Web Server)
- 정적 콘텐츠 제공: HTML, CSS, JavaScript, 이미지, 동영상 등 **정적 파일**을 클라이언트(브라우저)에게 전달합니다.
- HTTP 요청 처리: 클라이언트로부터 HTTP 요청을 받아 적절한 정적 리소스를 반환합니다.

WAS(Web Application Server)
- 동적 콘텐츠 제공: 데이터베이스 연동, 비즈니스 로직 처리, 세션 관리 등 동적인 웹 애플리케이션 기능을 제공합니다.
- 애플리케이션 실행 환경 제공: 웹 컨테이너를 제공하여 다양한 애플리케이션 컴포넌트를 실행할 수 있는 환경을 제공합니다.

**Comment by J-I-H-O** [comment]
- **Web Server**
    - HTTP 프로토콜을 기반으로 동작
    - 정적 콘텐츠(HTML, CSS, JS, 이미지 등) 제공
        - 단순 정적 콘텐츠를 제공할 때 Web Server를 사용하면 WAS를 이용하는 것 보다 빠르고 안정적으로 응답 가능
    - 동적 콘텐츠 요청 시 요청을 WAS로 전달
    - Apache, Nginx 등
- **WAS (Web Applicatioin Server)**
    - HTTP 프로토콜을 기반으로 동작
    - 웹 서버 기능을 포함하여 동적 리소스도 제공할 수 있는 Application 서버
        - 애플리케이션 로직을 실행하거나, DB를 조회하는 등의 동작 수행
    - 웹 서버 + 서블릿 컨테이너
    - Tomcat 등

---    

WAS는 Web Server의 기능을 포함하고 있기 때문에 WAS만으로도 시스템을 구성할 수 있습니다. 

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/1ed76e17-8255-414d-89f4-d4df7b6e4fa4)

그러나 이 경우, 정적 콘텐츠 응답의 역할과 애플리케이션 로직 실행 등의 역할을 WAS 혼자 수행하게 됩니다. 따라서 **서버 과부하의 위험**이 있을 수 있고, **정적 콘텐츠를 제공하느라 중요한 애플리케이션 로직이 지연될 수 있습니다.**
또한 WAS에 장애가 발생하는 경우, 정적 콘텐츠를 제공할 수 없게 되므로 **클라이언트에게 오류 화면을 전달할 수 없게 됩니다.**

따라서 다음과 같이 Web Server와 WAS를 분리해 사용합니다.

![image](https://github.com/2024-woowacourse-study/level-interview/assets/110461155/2dc2e5c9-9c65-407f-90ea-7768890018c0)

Web Server는 정적 콘텐츠를 처리하는데 집중하고, 동적인 처리가 필요한 경우 WAS에게 요청을 위임합니다. 이렇게 역할을 분리하여 과부화의 위험을 덜 수 있으며, WAS에 장애 발생 시에도 Web Server가 클라이언트에게 오류 화면을 전달할 수 있게 됩니다.

또한 Web Server와 WAS를 분리했기 때문에 개별적으로 리소스를 관리할 수 있습니다.
정적 리소스가 많이 사용된다면 Web Server를 증설하고, 
동적 리소스가 많이 사용된다면 WAS를 증설하고 Web Server가 로드 밸런싱 역할을 수행하게 하는 등, 효율적으로 관리할 수 있습니다. 
> 참고: 여러 대의 WAS를 사용하는 경우 fail over, fail back 처리에 유리합니다.

https://helloworld-88.tistory.com/71
[https://hstory0208.tistory.com/entry/Web과-WAS란-웹-서비스의-구조에-대해-알아보자](https://hstory0208.tistory.com/entry/Web%EA%B3%BC-WAS%EB%9E%80-%EC%9B%B9-%EC%84%9C%EB%B9%84%EC%8A%A4%EC%9D%98-%EA%B5%AC%EC%A1%B0%EC%97%90-%EB%8C%80%ED%95%B4-%EC%95%8C%EC%95%84%EB%B3%B4%EC%9E%90)

**Comment by le2sky** [comment]
로드 밸런싱 이미지 저장소

![image](https://github.com/user-attachments/assets/84507a0d-0526-43a1-850d-94cd6aaa3247)

![image](https://github.com/user-attachments/assets/41e4416f-e91c-4f04-8243-fa8bfa64eea4)

**Comment by le2sky** [comment]
<img width="658" alt="image" src="https://github.com/user-attachments/assets/e7de3c3b-7dd7-4bb7-a739-5eba0f035727" />

**Comment by le2sky** [comment]
![image](https://github.com/user-attachments/assets/d914948a-3d57-429c-9720-108d2c989cf5)
